codegen/masm/src/codegen.rs

```diff
@@ -0,0 +1,273 @@
+                let op = ir.constraint_graph().node(node_index).op();
+                self.visit_operation(op)
+            }
+        }
+    }
+
+    fn visit_operation(&mut self, op: &'ast Operation) -> Result<Self::Value, Self::Error> {
+        match op {
+            Operation::Value(value) => {
+                self.visit_value(value)?;
+            }
+            Operation::Add(left, right) => {
+                self.visit_node_index(left)?;
+                self.visit_node_index(right)?;
+                self.ins.push("add".into());
+            }
+            Operation::Sub(left, right) => {
+                self.visit_node_index(left)?;
+                self.visit_node_index(right)?;
+                self.ins.push("sub".into());
+            }
+            Operation::Mul(left, right) => {
+                self.visit_node_index(left)?;
+                self.visit_node_index(right)?;
+                self.ins.push("mul".into());
+            }
+            Operation::Exp(left, exp) => {
+                self.visit_node_index(left)?;
+                self.ins.push(format!("exp.{exp}").into());
+            }
+        };
+
+        Ok(())
+    }
+
+    fn visit_periodic_column(
+        &mut self,
+        _columns: &'ast PeriodicColumn,
+    ) -> Result<Self::Value, Self::Error> {
+        Ok(()) // TODO
+    }
+
+    fn visit_public_input(
+        &mut self,
+        _constant: &'ast PublicInput,
+    ) -> Result<Self::Value, Self::Error> {
+        Ok(())
+    }
+
+    fn visit_trace_access(
+        &mut self,
+        _trace_access: &'ast TraceAccess,
+    ) -> Result<Self::Value, Self::Error> {
+        Ok(()) // TODO
+    }
+
+    fn visit_value(&mut self, value: &'ast Value) -> Result<Self::Value, Self::Error> {
+        match value {
+            Value::BoundConstant(symbol) => match self.constants.entry(symbol.ident()) {
+                Entry::Occupied(entry) => match (entry.get(), symbol.access_type()) {
+                    (ConstantValueExpr::Scalar(scalar), AccessType::Default) => {
+                        self.ins.push(format!("push.{scalar}").into());
+                    }
+                    (ConstantValueExpr::Vector(vec), AccessType::Vector(pos)) => {
+                        let scalar = vec.get(*pos).ok_or(CodegenError::InvalidAccessType)?;
+                        self.ins.push(format!("push.{scalar}").into());
+                    }
+                    (ConstantValueExpr::Matrix(mat), AccessType::Matrix(x, y)) => {
+                        let scalar = mat
+                            .get(*x)
+                            .and_then(|v| v.get(*y))
+                            .ok_or(CodegenError::InvalidAccessType)?;
+                        self.ins.push(format!("push.{scalar}").into());
+                    }
+                    _ => return Err(CodegenError::InvalidAccessType),
+                },
+                Entry::Vacant(_) => return Err(CodegenError::UnknowConstant),
+            },
```

Comment:
One of the things we could do in the IR as an optimization pass is to get rid of all bound constants by converting them into inline constants. Once we do that, this could would not be necessary.